File: 3.x.x/crates-utopia/openapi/src/specification.rs
//! Canonical OpenAPI specification (PHP `Utopia\OpenAPI\Specification`).

use crate::json::Json;
use crate::model::{
    ExternalDocumentation, Info, Operation, PathItem, Schema, SecurityRequirement, SecurityScheme,
    Tag,
};
use crate::version::Version;
use indexmap::IndexMap;

/// Parsed OpenAPI document in the canonical model.
#[derive(Clone, Debug, PartialEq)]
pub struct Specification {
    pub version: Version,
    pub info: Info,
    pub servers: Vec<crate::model::Server>,
    pub tags: IndexMap<String, Tag>,
    pub paths: IndexMap<String, PathItem>,
    pub schemas: IndexMap<String, Schema>,
    pub security_schemes: IndexMap<String, SecurityScheme>,
    pub security: Vec<SecurityRequirement>,
    pub extensions: IndexMap<String, Json>,
    pub source_version: String,
    pub json_schema_dialect: Option<String>,
    pub external_documentation: Option<ExternalDocumentation>,
}

impl Specification {
    /// Flatten every operation across all paths (PHP `operations()`).
    pub fn operations(&self) -> Vec<&Operation> {
        let mut operations = Vec::new();
        for path in self.paths.values() {
            operations.extend(path.operations.values());
        }
        operations
    }

    /// Operations whose `tags` list contains `tag` (PHP `operationsByTag`).
    pub fn operations_by_tag(&self, tag: &str) -> Vec<&Operation> {
        self.operations()
            .into_iter()
            .filter(|op| op.tags.iter().any(|t| t == tag))
            .collect()
    }
}

File: packages/openapi/src/Specification.php
<?php

declare(strict_types=1);

namespace Utopia\OpenAPI;

use Utopia\OpenAPI\Model\ExternalDocumentation;
use Utopia\OpenAPI\Model\Info;
use Utopia\OpenAPI\Model\Operation;
use Utopia\OpenAPI\Model\PathItem;
use Utopia\OpenAPI\Model\ReferenceSchema;
use Utopia\OpenAPI\Model\Schema;
use Utopia\OpenAPI\Model\SecurityRequirement;
use Utopia\OpenAPI\Model\SecurityScheme;
use Utopia\OpenAPI\Model\Server;
use Utopia\OpenAPI\Model\Tag;

final readonly class Specification
{
    /**
     * @param  list<Server>  $servers
     * @param  array<string, Tag>  $tags
     * @param  array<string, PathItem>  $paths
     * @param  array<string, Schema>  $schemas
     * @param  array<string, SecurityScheme>  $securitySchemes
     * @param  list<SecurityRequirement>  $security
     * @param  array<string, mixed>  $extensions
     */
    public function __construct(
        public Version $version,
        public Info $info,
        public array $servers = [],
        public array $tags = [],
        public array $paths = [],
        public array $schemas = [],
        public array $securitySchemes = [],
        public array $security = [],
        public array $extensions = [],
        public string $sourceVersion = '',
        public ?string $jsonSchemaDialect = null,
        public ?ExternalDocumentation $externalDocumentation = null,
    ) {}

    /**
     * Resolve chained local component references without expanding recursive schema graphs.
     */
    public function resolveSchema(Schema $schema): Schema
    {
        $visited = [];
        while ($schema instanceof ReferenceSchema) {
            if (! str_starts_with($schema->reference, '#')) {
                break;
            }

            $name = null;
            $reference = rawurldecode($schema->reference);
            foreach (['#/components/schemas/', '#/definitions/'] as $prefix) {
                if (str_starts_with($reference, $prefix)) {
                    $name = str_replace(['~1', '~0'], ['/', '~'], substr($reference, \strlen($prefix)));
                    break;
                }
            }

            if ($name === null || isset($visited[$name]) || ! isset($this->schemas[$name])) {
                break;
            }
            $visited[$name] = true;
            $schema = $this->schemas[$name];
        }

        return $schema;
    }

    /** @return list<Operation> */
    public function operations(): array
    {
        $operations = [];
        foreach ($this->paths as $path) {
            array_push($operations, ...array_values($path->operations));
        }

        return $operations;
    }

    /** @return list<Operation> */
    public function operationsByTag(string $tag): array
    {
        return array_values(array_filter(
            $this->operations(),
            static fn(Operation $operation): bool => \in_array($tag, $operation->tags, true),
        ));
    }
}
